Banished Stmt types, added sequence exprs, lazy and eval operators

pencels opened the pull request:
- Statements are a thing of the past. Expressions are the one true AST type. This means things like `let` and `print` are now expressions:

    ```
    let x = (let y = 1; y + 1)
    x --> 2

    let z = (print "hello"; 5)
    -- "hello" printed
    z --> 5
    ```

- Added sequence exprs, delimited with the semicolon (`;`) operator. `;` has lower precedence than `=`, so you can assign values in sequences. Expressions are evaluated from left to right on sequence exprs. The last expr's value is the value of the sequence.

    ```
     -- Prints "hi", then evaluates to 2.
    print "hi"; 1 + 1
    ```

- Added "lazy" `&` and "eval" `*` operators, which will turn expressions into lazies or strictly eval a lazy value, respectively.

    ```
    let z = &print "Zzz" -- Equivalent to `{ print "Zzz" }`. Nothing prints here.
    *z -- Causes "Zzz" to print.
    ```

- Similar to the `{ x }` pattern, there is now a `&x` pattern which will strictly eval a lazy argument:

    ```
    -- Equivalent:
    { { x } -> ... }
    { &x -> ... }
    ```

- Added a distinction between simple strict eval (evaluate a lazy by returning the value of its inner expr) and full strict eval (evaluate a lazy by strictly evaluating its inner expr until the value is non-lazy). Simple strict eval is denoted by `*`, whereas full strict eval is denoted by `*!`:

    ```
    let nested = { print "hello"; { print "world" } }
    *nested -- Prints "hello" then returns { print "world" }
    *!nested -- Prints "hello" then "world"
    ```

- Patterns also carry the `*`/`*!` distinction. `&!x` or `!{ x }` will fully strictly evaluate the value matched to them.

    Strict patterns in matchboxes only try to strict-evaluate their arg values once. See [examples/lazy/patterns.slang](https://github.com/pencels/jslang/pull/6/commits/9c9fd265c0246970ad9f7859281d666d416c53b2#diff-679d9e28afe0e5bed74499916ec04d01) for examples.

- Redefined implicit strict eval within sequences/blocks. For a sequence with `n` exprs, the first `n - 1` exprs are fully strictly evaluated. The last expr is evaluated non-strictly. In other words `e1; e2; e3` is equivalent to `*!e1; *!e2; e3`:

    ```
    let effect = { print "hi" }
    effect; nothing -- Prints "hi", evaluates to nothing
    nothing; effect -- Evaluates to the { print "hi" } lazy value.
    ```

    To avoid a side-effect caused by implicit strict eval, assign the lazy to `_`, e.g. `_ = lazy; 4`.

- `()`-blocks allow sequencing of exprs like `{}`-blocks, but they strictly evaluate the sequence:

    ```
    let result = (
        print "hi"
        2
    ) -- "hi" gets printed here!

    result --> 2
    ```

- Opening a set of parens will also create a new environment, so names bound inside parens are not accessible outside their scope:

    ```
    (
        let x = 1
    )
    print x -- Name 'x' not defined here.
    ```

- Fixed some bugs with prefix operators not binding like prefix operators. Now `-1 + 2` parses correctly.

pencels committed:
Converted Stmts to Exprs, added sequence exprs

- Removed all references to Stmt, replaced with Expr.
- The semicolon (;) is not a binary operator with low precedence, used
  to sequence expressions.
- Block exprs and Lazies take only a single expression now, using
  sequence exprs to indicate sequences of "statements".
- Added parselets for Print and Let expressions. These have higher
  precedence than ';', meaning chaining like the following is allowed:
    let x = 2; print x; x * 3
- Fixed an issue with the PrefixOpParselet which allowed all other
  operators to bind tighter than prefix operators. e.g. -1;2 would
  parse as -(1;2) instead of (-1);2.
- Renamed SeqParselet to ListParselet as the old name is confusing.
- Some extra cleanup of unused files, e.g. UsedBindings.java.

pencels committed:
Add sequence exprs, lazy/eval operators

- ';' is an infix operator used to sequence expressions.
- () are the cousins of {}, they denote a block of exprs that are
  strictly evaluated rather than lazily evaluated.
- The & and * prefix operators will make a lazy value and evaluate
  lazies, respectively. * will fully evaluate nested lazies.
- The &x pattern is equivalent to the { x } pattern, it's used to eval
  lazy arguments to matchboxes.

pencels committed:
Added * and *! strict eval distinction

- Adds *!, which takes the place of the full strict eval operator. If
  its operand is a lazy, it will eval the lazy's body and return value
  until it reaches a non-lazy value.

- The * operator will do a single level of strict eval. If the return
  value of its argument is a lazy, it will not try to strictly evaluate
  it.

- Patterns have been extended to support these two levels of strict
  eval:
  - &x does a single level of strict eval on the argument it matches
  - { x } is equivalent to &x
  - &!x does full strict eval on the argument it matches
  - !{ x } is equivalent to &!x

- Removed implicit full strict eval (*!) at the program top-level.

- Edited prelude functions to support original behavior using the
  different * and *! operators.

- Added some more example files to demonstrate these changes.

- Added the '%' modulo operator just to use in examples.

compiler-errors reviewed (commented):
Left some thoughts. Let me know what you think of them.

pencels committed:
Changes to lazy eval semantics

- Restored the implicit full strict eval for exprs at the top-level and
  for exprs to the left of the `;` sequence operator, e.g. a; b; c is
  equivalent to  *!a; *!b; c.

- Assignment expressions return `nothing`. This allows "forgetting" of
  values, especially lazies, in sequences:

    _ = lazy_value; 5 -- Avoids *! evaluation of lazy_value in sequence.

- Added `if_unwrap` prelude fn which evaluates the chosen branch with *.

pencels committed:
Write ()-blocks so they spawn a new env

- This allows for local scoping of variables in strict expressions/blocks:

    let x = (let y = 0; y + 1)
    -- y is inaccessible here

pencels committed:
Fixed thunk handling for full/non-full patterns

- See examples/lazy/patterns.slang for examples.
